parser/lex.go

```diff
@@ -0,0 +1,605 @@
+// Based on design first introduced in: http://blog.golang.org/two-go-talks-lexical-scanning-in-go-and
+// Portions copied and modified from: https://github.com/golang/go/blob/master/src/text/template/parse/lex.go
+
+package parser
```

Comment:
I put the package-level comments in `parser.go`. The comments below the copyright here are for the file.